src/mutex.hpp

```diff
@@ -111,41 +120,110 @@ class CAPABILITY("mutex") Mutex {
 };
 
 
+// Defines an annotated interface for mutexes.
+// These methods can be implemented to use any internal mutex implementation.
+#include <shared_mutex>
+class CAPABILITY("mutex") MutexShared {
+private:
+	std::shared_timed_mutex std_mutex;
+public:
+  // Acquire/lock this mutex exclusively.  Only one thread can have exclusive
+  // access at any one time.  Write operations to guarded data require an
+  // exclusive lock.
+  void lock() ACQUIRE(){std_mutex.lock();}
+  // Try to acquire the mutex.  Returns true on success, and false on failure.
+  bool try_lock() TRY_ACQUIRE(true){return std_mutex.try_lock();}
+
+  // Acquire/lock this mutex for read operations, which require only a shared
+  // lock.  This assumes a multiple-reader, single writer semantics.  Multiple
+  // threads may acquire the mutex simultaneously as readers, but a writer
+  // must wait for all of them to release the mutex before it can acquire it
+  // exclusively.
+  void lock_shared() ACQUIRE_SHARED(){std_mutex.lock_shared();}
+  // Try to acquire the mutex for read operations.
+  bool try_lock_shared() TRY_ACQUIRE_SHARED(true) { return std_mutex.try_lock_shared(); }
+
+  // Release/unlock an exclusive mutex.
+  void unlock() RELEASE(){std_mutex.unlock();}
+
+  // Release/unlock a shared mutex.
+  void unlock_shared() RELEASE_SHARED(){std_mutex.unlock_shared();}
+
+  // Assert that this mutex is currently held by the calling thread.
+//  void AssertHeld() ASSERT_CAPABILITY(this);
+
+  // Assert that is mutex is currently held for read operations.
+//  void AssertReaderHeld() ASSERT_SHARED_CAPABILITY(this);
+
+  // For negative capabilities.
+//  const Mutex& operator!() const { return *this; }
+};
+
+
 // MutexLocker is an RAII class that acquires a mutex in its constructor, and
 // releases it in its destructor.
 template <class t_mutex>
-class SCOPED_CAPABILITY Lock_guard {
+class SCOPED_CAPABILITY LockGuard {
 private:
   t_mutex &mut;
 
 public:
-  Lock_guard(t_mutex &mu) ACQUIRE(mu) : mut(mu) {
-    mut.lock();
-  }
-  ~Lock_guard() RELEASE() {
-    mut.unlock();
-  }
+  LockGuard(t_mutex &mu) ACQUIRE(mu) : mut(mu) { mut.lock(); }
+  ~LockGuard() RELEASE() { mut.unlock(); }
+  void lock() ACQUIRE() { mut.lock(); }
+  void unlock() RELEASE() { mut.unlock();  }
 };
 
 template <class t_mutex>
-class SCOPED_CAPABILITY Unique_lock {
+class SCOPED_CAPABILITY UniqueLockGuardRO {
 private:
-  std::unique_lock<t_mutex> unique_lock;
-
+  t_mutex & m_mut;
+  bool m_locked; ///< not all mutex types know if they are locked apparentyl?
```

Comment:
(discussed)
we can do that, it's not data race, it's in Unique lock (not in mutex)
so we can do that by hand, and it's safe